controllers/key_controller.go

```diff
@@ -0,0 +1,274 @@
+			return ctrl.Result{RequeueAfter: requeueAfter}, nil
+		}
+		log.Error(err, "unexpected error fetching referenced repository")
+		return ctrl.Result{RequeueAfter: requeueAfter}, err
+		// TODO(stealthybox): This requeue period causes the reconcile to poll for the RepositoryRef.
+		// 	It would be an improvement to implement it with exponential backoff, potentially on a Condition similar to Pods failing on a required SecretRef.
+		// 	Ideally, we should add a Watch for this particular Repository to a map somewhere or check all of the Keys and queue matching ones when a Repository is queued.
+		// 	There are some examples around the internet of subscribing to related objects in a controller.
+		// 	This can't be implemented with non-controller, non-blocking ownerRefs because the Repository will still be unintentionally deleted in the background after deleting the Key.
+		// 	A re-sync masks this issue.
+	}
+
+	// wait for the referenced repository to sync
+	if repository.Status.Status != v1alpha1.SyncedStatus {
+		r.updateKeyStatus(ctx, &key, v1alpha1.WaitingStatus)
+		log.Info("referenced repository not yet synced")
+		return ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
+
+	// Repo and Secret are both ready, add finalizer for github-Delete before doing github-Create
+	if key.ObjectMeta.DeletionTimestamp.IsZero() &&
+		!containsString(key.GetFinalizers(), keyFinalizerName) {
+		log.Info("adding finalizer", "name", key.Name)
+		if err := r.addFinalizer(ctx, &key); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	var ghKey *github.Key
+	var resp *github.Response
+	var err error
+
+	createGHKeyAndUpdate := func() (ctrl.Result, error) {
+		r.updateKeyStatus(ctx, &key, v1alpha1.CreatingStatus)
+		if ghKey, err = r.GitClient.CreateKey(ctx, repository.Spec.Organization, repository.GetName(), &key, &secret); err != nil {
+			return ctrl.Result{RequeueAfter: requeueAfter}, err
+		}
+
+		err = r.updateKeyStatusDetails(ctx, &repository, ghKey, &key)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: requeueAfter}, err
+		}
+		return ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
+
+	if key.Status.GitHubKeyID == 0 {
+		log.Info("creating new key in GitHub")
+		return createGHKeyAndUpdate()
+	}
+
+	ghKey, resp, err = r.GitClient.GetKey(ctx, repository.Spec.Organization, repository.GetName(), key.Status.GitHubKeyID)
+	if err != nil && isNotFound(resp) {
+		log.Info("expected key not found, creating new key in GitHub", "missingID", key.Status.GitHubKeyID)
+		return createGHKeyAndUpdate()
+		// note: this can occur on a re-sync despite there being no watch on the GitHub API objects
+	} else if err != nil {
+		log.Error(err, "error fetching key from GitHub")
+		return ctrl.Result{RequeueAfter: requeueAfter}, err
+	}
+
+	// recreateGHKey indicates whether the key in GitHub does not match the Key object declaration
+	recreateGHKey := (strings.TrimSpace(ghKey.GetKey()) != strings.TrimSpace(key.Status.PublicKey) ||
```

Comment:
Does this get weird if you register a key with an email at the end, doesn't Github truncate that… Like if the key was:

```
...hx38Ilh4f chris@chrishein.local
```

would this still get caught?